[feature] init template

Soecka opened the pull request:
[add] github actions, husky, lint-staged, pull request template
[add] shadcn/ui, tailwindcss, eslint, prettier
[add] i18n

Soecka committed:
[add] github actions, husky, lint-staged, pull request template, issue templates
[add] shadcn/ui, tailwindcss, eslint, prettier

Soecka committed:
[add] i18n route

TechQuery commented on postcss.config.mjs:
```diff
@@ -0,0 +1,5 @@
+const config = {
+  plugins: ["@tailwindcss/postcss"],
+};
+
+export default config;
```
```suggestion
export default {
  plugins: ["@tailwindcss/postcss"],
};
```

Soecka replied:
改成建议形式会有 warning，eslint 这个规则默认设置的是 warning，next-create-app 跑出来也是这个形式，建议保留不更改或考虑配置对应 eslint 规则为不默认导出视为 error

https://github.com/import-js/eslint-plugin-import/blob/v2.31.0/docs/rules/no-anonymous-default-export.md

TechQuery replied:
> 考虑配置对应 eslint 规则为不默认导出视为 error

需要类型标注时不能用默认导出，暂时这样不改了。

TechQuery commented on lib/utils.ts:
```diff
@@ -0,0 +1,6 @@
+
+export function cn(...inputs: ClassValue[]) {
+  return twMerge(clsx(inputs));
+}
```
```suggestion
export const cn = (...inputs: ClassValue[]) => twMerge(clsx(inputs));
```

TechQuery commented on .github/workflows/init-template.yml:
```diff
@@ -0,0 +1,35 @@
+      - main
+env:
+  TEMPLATE_OWNER: idea2app
+  TEMPLATE_NAME: Next-Bootstrap-ts
```
```suggestion
  TEMPLATE_NAME: Next-shadcn-ts
```

TechQuery commented on README.md:
```diff
@@ -1 +1,36 @@
-# Next-shadcn-ts
```
一级标题保留

TechQuery commented on README.md:
```diff
@@ -1 +1,36 @@
-# Next-shadcn-ts
\ No newline at end of file
+This is a [Next.js](https://nextjs.org) project bootstrapped with [`create-next-app`](https://nextjs.org/docs/app/api-reference/cli/create-next-app).
```
文中链接用 Git Pager 刷一遍。

Soecka replied:
https://github.com/DavidAnson/markdownlint/blob/v0.37.4/doc/md053.md

Soecka replied:
@TechQuery  下面这些 lint 插件和规则需要加吗

TechQuery replied:
> @TechQuery  下面这些 lint 插件和规则需要加吗

加

TechQuery commented on app/[lang]/layout.tsx:
```diff
@@ -0,0 +1,36 @@
+
+export default function RootLayout({
+  children,
+}: Readonly<{
+  children: ReactNode;
+}>) {
```
```suggestion
}: PropsWithChildren<{}>) {
```

Soecka replied:
加插件规则？

```
// ... existing code ...
      "@typescript-eslint/consistent-type-definitions": ["error", "interface"],
      "no-restricted-syntax": [
        "error",
        {
          "selector": "TSPropertySignature[key.name='children']",
          "message": "Please use PropsWithChildren<T> instead of defining children manually"
        }
      ],
// ... existing code ...
```

TechQuery replied:
> 加插件规则？
> 
> ```
> // ... existing code ...
>       "@typescript-eslint/consistent-type-definitions": ["error", "interface"],
>       "no-restricted-syntax": [
>         "error",
>         {
>           "selector": "TSPropertySignature[key.name='children']",
>           "message": "Please use PropsWithChildren<T> instead of defining children manually"
>         }
>       ],
> // ... existing code ...
> ```

还可以这样？加

TechQuery commented on app/[lang]/page.tsx:
```diff
@@ -0,0 +1,30 @@
+  params: Promise<{ lang: Locale }>;
+}) {
+  const lang = (await params).lang;
+  const dict = await getDictionary(lang);
+  return (
```
```suggestion
  const dict = await getDictionary(lang);

  return (
```

Soecka replied:
https://eslint.style/rules/js/padding-line-between-statements

TechQuery commented on app/api/og/route.tsx:
```diff
@@ -0,0 +1,29 @@
+import { siteName } from "@/config";
+
+export async function GET() {
+  return new ImageResponse(
+    (
+      <div
```
图片响应能返回 `<div />`？

Soecka replied:
https://nextjs.org/docs/app/building-your-application/optimizing/metadata#dynamic-image-generation

TechQuery commented on app/sitemap.ts:
```diff
@@ -0,0 +1,12 @@
+import { baseURL } from "@/config";
+
+export default function sitemap(): MetadataRoute.Sitemap {
+  return [
+    {
+      url: baseURL,
+      lastModified: new Date(),
+    },
+  ];
```
```suggestion
  return [{ url: baseURL, lastModified: new Date() }];
```

TechQuery commented on eslint.config.mjs:
```diff
@@ -0,0 +1,68 @@
+
+const compat = new FlatCompat({ baseDirectory });
+
+const eslintConfig = [
```
```suggestion
export default [
```

TechQuery commented on get-dictionary.ts:
```diff
@@ -0,0 +1,16 @@
+// We enumerate all dictionaries here for better linting and typescript support
+// We also get the default import for cleaner types
+const dictionaries = {
+  en: () => import("./dictionaries/en.json").then((module) => module.default),
```
```suggestion
  en: () => import("./dictionaries/en.json").then(({ default }) => default),
```

Soecka replied:
```
      "prefer-destructuring": ["error", {
        "object": true,
        "array": true
      }],
```

TechQuery commented on middleware.ts:
```diff
@@ -0,0 +1,58 @@
+
+import { i18n } from "./i18n-config";
+
+function getLocale(request: NextRequest): string | undefined {
+  // Negotiator expects plain object so we need to transform headers
+  const negotiatorHeaders: Record<string, string> = {};
+  request.headers.forEach((value, key) => (negotiatorHeaders[key] = value));
+
+  const { locales } = i18n;
+
+  // Use negotiator and intl-localematcher to get best locale
+  const languages = new Negotiator({ headers: negotiatorHeaders }).languages([
+    ...locales,
+  ]);
+
+  const locale = matchLocale(languages, locales, i18n.defaultLocale);
+
+  return locale;
```
```suggestion
function getLocale({ headers }: NextRequest): string | undefined {
  const { defaultLocale, locales } = i18n;

  // Use negotiator and intl-localematcher to get best locale
  const languages = new Negotiator({ headers: Object.fromEntries([...headers]) }).languages(locales);

  return matchLocale(languages, locales, defaultLocale);
```

TechQuery commented on middleware.ts:
```diff
@@ -0,0 +1,58 @@
+  // Redirect if there is no locale
+  if (pathnameIsMissingLocale) {
+    const locale = getLocale(request);
+    console.log(
+      locale,
+      "======redirect",
+      pathnameIsMissingLocale,
+      "=======pathnameIsMissingLocale",
+    );
```
```suggestion
```

Soecka replied:
```

    "no-console": ["error", { 
        "allow": ["warn", "error", "info"] 
      }],
```

TechQuery commented on tsconfig.json:
```diff
@@ -0,0 +1,27 @@
+{
+  "compilerOptions": {
+    "target": "ES2017",
+    "lib": ["dom", "dom.iterable", "esnext"],
+    "allowJs": true,
+    "skipLibCheck": true,
+    "strict": true,
+    "noEmit": true,
+    "esModuleInterop": true,
+    "module": "esnext",
```
```suggestion
    "lib": ["DOM", "DOM.Iterable", "ESNext"],
    "allowJs": true,
    "skipLibCheck": true,
    "strict": true,
    "noEmit": true,
    "esModuleInterop": true,
    "module": "ESNext",
```

TechQuery commented on package.json:
```diff
@@ -0,0 +1,57 @@
+    "class-variance-authority": "^0.7.1",
+    "clsx": "^2.1.1",
+    "lucide-react": "^0.475.0",
+    "next": "15.1.7",
```
```suggestion
    "next": "^15.1.7",
```

TechQuery commented on package.json:
```diff
@@ -0,0 +1,57 @@
+    "husky": "^9.1.7",
+    "lint-staged": "^15.4.3",
+    "tailwindcss": "^4.0.8",
+    "typescript": "^5.7.3"
```
```suggestion
    "typescript": "~5.7.3"
```

TechQuery commented on package.json:
```diff
@@ -0,0 +1,57 @@
+    "dev": "next dev --turbopack",
+    "build": "next build",
+    "start": "next start",
+    "lint": "next lint",
+    "lint:all": "eslint --fix .",
```
```suggestion
    "lint": "next lint --fix",
```
Next.js 封装的命令就是处理全部。

Soecka committed:
[optimize] use flat config for eslint
[add] stylelint plugin
[remove] unused svg icons
[polish] opengraph image route

TechQuery commented on app/[lang]/page.tsx:
```diff
@@ -1,25 +1,35 @@
+export const metadata: Metadata = {
+  title: siteName,
+  description: description,
+  authors: [{ name: author }],
```
```suggestion
  authors: { name: author }[],
```

TechQuery commented on app/api/og/route.tsx:
```diff
@@ -1,24 +1,82 @@
+          style={{
+            display: "flex",
+            alignItems: "center",
+            justifyContent: "center",
```
前端用的是 Tailwind CSS，后端返回也可以直接用它的工具类。

Soecka replied:
tailwind 支持是 experimental ，先不建议使用

TechQuery replied:
> tailwind 支持是 experimental ，先不建议使用

CSS 库支持为何还需要”试验“？整个项目都在用啊，只要是 JSX 就可以写对应类名啊。

TechQuery commented on app/api/og/route.tsx:
```diff
@@ -1,24 +1,82 @@
+              d="M73 73L136.22 109.5V182.5L73 219L9.78014 182.5V109.5L73 73Z"
+              fill="#21A9A8"
+              fill-opacity="0.7"
+            ></path>
```
```suggestion
            />
```

Soecka replied:
/* eslint-disable react/self-closing-comp */

包含 svg 的文件一般会关闭这条规则，要不然要改的东西太多了

TechQuery replied:
> /* eslint-disable react/self-closing-comp */
> 
> 包含 svg 的文件一般会关闭这条规则，要不然要改的东西太多了

ESLint 不能 `--fix` 它？

TechQuery commented on app/api/og/route.tsx:
```diff
@@ -1,24 +1,82 @@
+            justifyContent: "center",
+          }}
+        >
+          <svg
+            focusable="false"
+            aria-hidden="true"
+            viewBox="0 0 234 219"
+            width="234"
+            height="219"
+            fill="none"
+            xmlns="http://www.w3.org/2000/svg"
+          >
```
SVG 由独立模块导入。

Soecka replied:
导入不了，只能通过相对 URL 转 base64，edge 运行时有较多 API 不支持

TechQuery commented on app/api/og/route.tsx:
```diff
@@ -1,24 +1,82 @@
+          style={{
+            display: "flex",
+            fontSize: 40,
+            fontStyle: "normal",
+            color: "black",
+            marginTop: 30,
+            lineHeight: 1.8,
+            whiteSpace: "pre-wrap",
+          }}
+        >
+          <b>{siteName}</b>
```
```suggestion
            color: "black",
            marginTop: 30,
            lineHeight: 1.8,
            whiteSpace: "pre-wrap",
          }}
        >
          <strong>{siteName}</strong>
```
从语义上应该用 `<figure />` 系列标签。

Soecka replied:
这个是用来生成图片的，不支持这个标签
https://github.com/vercel/satori/blob/main/src/handler/presets.ts

TechQuery commented on config.ts:
```diff
@@ -1,3 +1,5 @@
 export const baseURL = "https://idea2.app";
-export const siteName = "idea2app";
+export const siteName = "Next-shadcn-ts";
 export const description = "Make an Idea to an App";
```
描述应该是脚手架的描述。

Soecka replied:
多加 idea2app 链接引流，那 baseURL 和 description 要不要改

TechQuery replied:
> 多加 idea2app 链接引流，那 baseURL 和 description 要不要改

好吧，先这样吧。

TechQuery commented on package.json:
```diff
@@ -4,35 +4,42 @@
     "eslint": "^9.21.0",
-    "eslint-config-next": "15.1.7",
-    "eslint-config-prettier": "^10.0.1",
+    "eslint-config-next": "15.2.1",
```
```suggestion
    "eslint-config-next": "^15.2.1",
```

TechQuery commented on package.json:
```diff
@@ -4,35 +4,42 @@
+    "@stylistic/eslint-plugin": "^4.2.0",
+    "@tailwindcss/postcss": "^4.0.10",
+    "@types/eslint-config-prettier": "^6.11.3",
+    "@types/negotiator": "0.6.3",
```
```suggestion
    "@types/negotiator": "^0.6.3",
```

Soecka replied:
https://github.com/JoshuaKGoldberg/eslint-plugin-package-json

https://github.com/matzkoh/prettier-plugin-packagejson

可以给 package json package 排序和检查语义化版本

TechQuery replied:
> 可以给 package json package 排序和检查语义化版本

键名字母排序不需要，一般按照”元数据、Git 仓库、依赖包、第三方配置、脚本“的语义顺序排列。

版本检查可以尝试配置。

TechQuery commented on eslint.config.mjs:
相同建议：https://github.com/idea2app/Next-Bootstrap-ts/pull/50#pullrequestreview-2663940185

Soecka commented on app/api/og/route.tsx:
```diff
@@ -1,24 +1,82 @@
 import { siteName } from "@/config";
 
+/**
+ * @see {@link https://og-playground.vercel.app/}
```
https://og-playground.vercel.app/

目前支持的类型预览

Soecka committed:
[add] dynamic metadata
[polish] i18n structure
[optimize] open graph image

TechQuery commented on app/api/og/route.tsx:
```diff
@@ -0,0 +1,69 @@
+          }}
+        >
+          <img src={logoDataUrl} width={200} height={200} />
+          <hgroup tw="flex flex-col items-center" style={{ gap: "16px" }}>
+            <h2 tw="font-bold text-black text-4xl m-0">{title}</h2>
+            <p tw="text-gray-500 text-lg">{description}</p>
+          </hgroup>
+        </div>
+      ),
+      {
+        width: 1200,
+        height: 630,
+      },
```
```suggestion
          <hgroup tw="flex flex-col items-center" style={{ gap: "1rem" }}>
            <h2 tw="font-bold text-black text-4xl m-0">{title}</h2>
            <p tw="text-gray-500 text-lg">{description}</p>
          </hgroup>
        </div>
      ),
      { width: 1200, height: 630 },
```

TechQuery commented on app/api/og/route.tsx:
```diff
@@ -0,0 +1,69 @@
+  } catch (error) {
+    console.error("OG Image generation error:", error);
+
+    return new Response(`Failed to generate image: ${error}`, { status: 500 });
```
```suggestion
    return new Response(`Failed to generate image: ${error.message}`, { status: 500 });
```
